### ISSUE TITLE
Encapsulate unsafe functions in Unsafe submodules?

### ISSUE THREAD

**Issue by scolsen** [carp-lang/Carp#686]
I forget whether or not we discussed this approach when we decided to ultimately prefix unsafe functions with `unsafe-`--but I'm wondering if it would save us some keystrokes and make more sense to add nested Unsafe modules to house such functions? 

That way, to call an unsafe func one needs to write:

```clojure
(Array.Unsafe.nth xs 1)
```

And the safe variant is:

```clojure
(Array.nth xs 1)
```

Plus, one can decide to use unsafe variants wholecloth in certain contexts via `(use Foo.Unsafe)` etc.

Related: #513

**Comment by hellerve** [comment]
We did talk about this in the chat, but it’s good that you opened an issue for this; this way we won’t lose track of it :)

**Comment by eriksvedang** [comment]
I think this is a very interesting idea and would like to move in that direction. I have a feeling it might have some usability repercussions in other places so we should try it out properly before commiting to the idea. If anyone wants to do a proof of concept in a branch I'm happy to try it out!

**Comment by scolsen** [comment]
> We did talk about this in the chat, but it’s good that you opened an issue for this; this way we won’t lose track of it :)

haha, I had a vague memory of it but wasn't sure! 😅

**Comment by eriksvedang** [comment]
This would actually clash with making `nth` and some other array functions into interfaces, which we're on track to do now that the higher kinded types work. So maybe it's not such a good idea in the end? (though I like the aesthetics of it in the examples above)

**Comment by hellerve** [comment]
Yes, that actually seems like it would break quite a few assumptions, which is a bummer. I’ll definitely investigate whether this is still feasible, though!

**Comment by scolsen** [comment]
Ahhh great point! @eriksvedang yeah, it would be pretty unfortunate to call `nth` generically and have it resolve to the unsafe variant! Actually, as interfaces currently work, we couldn't even do this since the types would clash and there'd be two possible implementations of `nth` that work on `Array`

One way to get around this is to just move safety to the type level. So, we'd have two Arrays,

- `Array` which admits of only safe operations
- `UnsafeArray` which admits of only unsafe operations

Then the implementations of `nth` would have different signatures:

`interface nth (Fn [(Ref (f a)) Int] a)`
- `Array.nth (Fn [(Ref (Array a)) Int] (Maybe a))`
- `UnsafeArray.nth (Fn [(Ref (UnsafeArray a)) Int] a)`

Then, in order to make it easy to use unsafe methods flexibly on array's that are original safe (much like we n=use them now) we can either introduce one or two new functions:

`unsafe (Array a) -> (UnsafeArray a)` and `safe (UnsafeArray a) -> (Array a)`

Or we just assume that everyone immediately wants to return to safety and only provide a single function: `unsafe :: (Array a) -> (Fn [(UnsafeArray a)] (UnsafeArray a)) -> (Array a)`

A higher order that takes an array, takes some unsafe operation (possibly a sequence of several) performs them on the array, then converts it back to a safe type, the only rule being that the `UnsafeArray` function it takes as an argument *must* return an `UnsafeArray` (it must be functorial)--then the body of the function can perform whatever manipulations the user needs. The only issue with this approach is that it makes things like the unsafe variant of `nth` a tad more difficult to use (since one has to return an `UnsafeArray` but not impossible.

Effectively, it'd be roughly equivalent to moving all unsafe operations into an "unsafe monad"--a special context in which they can be performed, but one always gets back to safety.

**Comment by scolsen** [comment]
> One way to get around this is to just move safety to the type level. So, we'd have two Arrays,

I should also note that some other languages do this to some extent, though the differences are not always oriented around "safety" explicitly. For example, Standard ML has two types, `Array` and `Vector`. Array is the type of mutable sequences of values while `Vector` is the type of immutable sequences of values. One isn't really safer than the other (since `Vector` also permits of an `nth` operation that can raise an exception), but their mutability differences do mean they have different semantics and admit of different operations, just like the safe/unsafe cases, and it does mean that making similar distinctions at the type level isn't unheard of. 

Haskell too, sort of has distinctions like this, e.g. `List.NonEmpty` vs. `List`.
